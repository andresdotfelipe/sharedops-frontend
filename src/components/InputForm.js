import React from 'react';
import { Col, Form } from 'react-bootstrap';

const adaptFileEventToValue = handler =>  async ({ target: { files } }) => {    
    try {
        if (files[0]) {
            const header = await getFileHeader(files[0]);
            handler({ file: files[0], name: files[0].name, header });
        }
    } catch (e) {
        console.warn(e.message)
        handler(null);
    }        
};

const getFileHeader = blob => {    
    let fileReader = new FileReader();
    return new Promise((resolve, reject) => {    
        fileReader.onloadend = e => {
            /*                
                The array buffer created by the FileReader instance
                is a generic way to represent a binary buffer. So
                a TypedArray is necessary, in this case, a
                a Uint8Array with the first 4 elements. 
            */
            let arr = (new Uint8Array(e.target.result)).subarray(0, 4);
            let header = '';            
            //  Retrieves every byte and transforms it to hexadecimal.            
            for (let i = 0; i < arr.length; i++) {
                header += arr[i].toString(16);
            }
            resolve(header);
        };
        fileReader.readAsArrayBuffer(blob);
    });        
};

const InputForm = ({
    meta: { error, touched },
    input: { name, value, onChange, onBlur },
    label, type, ph, as, rows, disabled
}) => {                
    
    return (
        <>  
            <Form.Group as={Col} md="12">
                {
                    label &&
                    <Form.Label>
                        {label}
                    </Form.Label>
                }                                                                                                      
                {
                    type === 'file' ?
                    <Form.Control
                        isInvalid={error && touched}
                        isValid={!error && touched}
                        name={name} 
                        type={type}                            
                        onChange={adaptFileEventToValue(onChange)}        
                        onBlur={adaptFileEventToValue(onBlur)}
                    /> :
                    <Form.Control
                        isInvalid={error && touched}
                        isValid={!error && touched}
                        name={name} 
                        type={type}
                        value={value}
                        placeholder={ph}
                        as={as}
                        rows={rows}
                        disabled={disabled}
                        onChange={onChange}        
                        onBlur={onBlur}
                    />     
                }                                                                   
                <Form.Control.Feedback type="invalid">
                    {error}
                </Form.Control.Feedback>  
            </Form.Group>
        </>
    );
};

export default InputForm;