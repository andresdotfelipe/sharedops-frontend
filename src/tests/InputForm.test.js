import React from 'react';
import { shallow } from 'enzyme';
import InputForm from '../components/InputForm';

const onChange = jest.fn();
const adaptFileEventToValue = jest.fn();
const defaultProps = {
    meta: {
        error: '',
        touched: false
    },
    input: {
        name: 'name',        
        onChange,
        onBlur: jest.fn()
    },
    label: 'label',                
    disabled: false
};

describe('InputForm component', () => {
    
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('updates input text value when changed', () => {
        const inputTextProps = {
            ...defaultProps, 
            type: 'text',
            ph: 'placeholder'            
        };
        const wrapper = shallow(<InputForm {...inputTextProps} />);
        const event = { target: { value: 'Test value' } };
        wrapper.find('FormControl').simulate('change', event);
        expect(onChange).toHaveBeenCalledWith(event);
    });

    it('renders correctly when input is textarea', () => {
        const inputTextAreaProps = {
            ...defaultProps, 
            as: 'textarea',
            rows: 10,
            ph: 'placeholder'            
        };
        const wrapper = shallow(<InputForm {...inputTextAreaProps} />);        
        const event = { target: { value: 'Test value' } };
        wrapper.find('FormControl').simulate('change', event);
        expect(onChange).toHaveBeenCalledWith(event);
        
    });

    it('renders correctly when input is file', () => {
        try {            
            defaultProps.onChange = adaptFileEventToValue(onChange);        
            const inputFileProps = {
                ...defaultProps, 
                type: 'file'            
            };
            const wrapper = shallow(<InputForm {...inputFileProps} />);            
            const file =  new File([new ArrayBuffer(1)], 'file.jpg');        
            const event = { target: { files: { 0: file, length: 1 } } };         
            wrapper.find('FormControl').simulate('change', event);
            expect(adaptFileEventToValue).toHaveBeenCalled();
        } catch (e) {
            console.log(e);
        }        
    });
});