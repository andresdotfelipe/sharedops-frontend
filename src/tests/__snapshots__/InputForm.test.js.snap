// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm component renders correctly when input is file 1`] = `
<div
  className="form-group col-md-12"
>
  <label
    className="form-label"
  >
    label
  </label>
  <input
    className="form-control-file"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    type="file"
  />
  <div
    className="invalid-feedback"
  />
</div>
`;

exports[`InputForm component renders correctly when input is text 1`] = `
<div
  className="form-group col-md-12"
>
  <label
    className="form-label"
  >
    label
  </label>
  <input
    className="form-control"
    disabled={false}
    name="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    placeholder="placeholder"
    type="text"
    value="value"
  />
  <div
    className="invalid-feedback"
  />
</div>
`;

exports[`InputForm component renders correctly when input is textarea 1`] = `
<div
  className="form-group col-md-12"
>
  <label
    className="form-label"
  >
    label
  </label>
  <textarea
    className="form-control"
    disabled={false}
    name="name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    placeholder="placeholder"
    rows={10}
    value="value"
  />
  <div
    className="invalid-feedback"
  />
</div>
`;
